xen/arch/arm/coproc/coproc.c

```diff
@@ -437,6 +437,36 @@ void coproc_release(struct coproc_device *coproc)
     xfree(coproc);
 }
 
+void __iomem *coproc_map_offset(struct coproc_device *coproc, u32 mmio_idx,
+                                u64 offset, u64 size)
+{
+    void __iomem *base;
+
+    if ( mmio_idx > coproc->num_mmios )
+        return ERR_PTR(-EINVAL);
+
+    if ( offset + size > coproc->mmios[mmio_idx].size )
+        return ERR_PTR(-EINVAL);
+
+    /* if offset is not on page boundary then align */
+    base = ioremap_nocache(coproc->mmios[mmio_idx].addr + (offset & PAGE_MASK),
+                           size);
```

Comment:
Here you should also keep tracking of crossing page boundaries by offset+size, so you could need to map a one page more.

Reply:
ioremap_nocache takes care of it internally (which is expected):
    /* allow non page aligned addresses but for mapping we need to align them */
    offset = (phys_addr & ~PAGE_MASK);
    num_pages = (offset + size + PAGE_SIZE - 1) / PAGE_SIZE;